Simplify non-scaling-stroke handling in (Legacy)RenderSVGShape

heycam opened the pull request:
#### b8abd5adb45e7aa2f3c232740d232e3b29c594dc
<pre>
Simplify non-scaling-stroke handling in (Legacy)RenderSVGShape
<a href="https://bugs.webkit.org/show_bug.cgi?id=241754">https://bugs.webkit.org/show_bug.cgi?id=241754</a>

Reviewed by NOBODY (OOPS!).

(Legacy)RenderSVGShape::nonScalingStrokePath has a static variable that
stores the returned Path, presumably so that it can live long enough for
the caller to make use of it. We can simplify this by returning Path
object directly.

While we&apos;re here, we can also simplify some other functions dealing with
the non-scaling stroke transform and usage of the non-scaling stroke
path.

We add a new AffineTransform::mapPath function, which does what the
current nonScalingStrokePath function does.

* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::Path::transform):
* Source/WebCore/platform/graphics/transforms/AffineTransform.cpp:
(WebCore::AffineTransform::mapPath const):
* Source/WebCore/platform/graphics/transforms/AffineTransform.h:
* Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::strokeShape const):
(WebCore::LegacyRenderSVGShape::shapeDependentStrokeContains):
(WebCore::LegacyRenderSVGShape::setupNonScalingStrokeContext):
(WebCore::LegacyRenderSVGShape::strokeShape):
(WebCore::LegacyRenderSVGShape::calculateStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::nonScalingStrokePath const): Deleted.
* Source/WebCore/rendering/svg/LegacyRenderSVGShape.h:
(WebCore::LegacyRenderSVGShape::strokeStyleApplier const):
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::strokeShape const):
(WebCore::RenderSVGPath::zeroLengthLinecapPath const):
* Source/WebCore/rendering/svg/RenderSVGPath.h:
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::strokeShape const):
(WebCore::RenderSVGShape::shapeDependentStrokeContains):
(WebCore::RenderSVGShape::setupNonScalingStrokeContext):
(WebCore::RenderSVGShape::strokeShape):
(WebCore::RenderSVGShape::calculateStrokeBoundingBox const):
(WebCore::RenderSVGShape::nonScalingStrokePath const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGShape.h:
(WebCore::RenderSVGShape::strokeStyleApplier const):
</pre>

nikolaszimmermann commented on Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
```diff
@@ -101,8 +100,8 @@ bool LegacyRenderSVGShape::shapeDependentStrokeContains(const FloatPoint& point,
         AffineTransform nonScalingTransform = nonScalingStrokeTransform();
-        Path* usePath = nonScalingStrokePath(m_path.get(), nonScalingTransform);
-        return usePath->strokeContains(nonScalingTransform.mapPoint(point), [this] (GraphicsContext& context) {
+        Path usePath = nonScalingTransform.mapPath(path());
```
s/Path/auto/ - no gained information by adding the Path return type here.

nikolaszimmermann commented on Source/WebCore/platform/graphics/transforms/AffineTransform.cpp:
```diff
@@ -349,6 +350,13 @@ Region AffineTransform::mapRegion(const Region& region) const
 
+Path AffineTransform::mapPath(const Path& path) const
+{
+    Path transformedPath = path;
```
I would expect a dedicated case here (similar to mapRegion / ...) to handle at least identity matrices, or even better also translations. Obviously no benefit for non-scaling-stroke transform, but mapPath() should be generic, if exposed like this.

```
if (isIdentityOrTranslation()) {
    Path mappedPath(path);
    mappedPath.translate(FloatSize(narrowPrecisionToFloat(m_transform[4], ...
    return mappedPath;
}
``` 

Note: For CG there is no actual change, since Path::translate() does construct a temporary AffineTransform to use Path::transform() internally. However for Cairo Path::translate() is specifically optimized, to actually translate all path consituents, since Cairos Path::transform() logic is much slower than CoreGraphics.

heycam replied:
What do you think about making Cairo's Path::translate and Path::transform check for an identity transform, rather than having that check in AffineTransform::mapPath? The CG version of Path::transform does this (and Path::translate just calls transform).

shallawa replied:
I am a little bit confused by `AffineTransform::mapPath()` and `Path::transform()`. Their names do not show the difference between them. But by inspecting the code, I think `AffineTransform::mapPath()` will do the transformation out-of-place while `Path::transform()` will do it in-place. But `Path::transform()` reallocates all the Path internal data. So there is no much difference in memory between them. Can we get rid of `Path::transform()`? Do we really need it? Does it save us anything?

nikolaszimmermann replied:
I actually expect "mapPath" to not alter the Path by the transformation ("out-of-place"), whereas Path::tranform() indicates in-place operations to me. I don't find it too confusing, but also wouldn't mind a more friendly naming scheme, if present.

Furthermore you only looked at the CG variant of Path::transform(), which indeed reallocates (CGPathCreateMutableCopyByTransformingPath), as that's the preferred CG solution. However the Cairo variant in PathCairo.cpp, actually keeps the internal path data and transforms each coordinate. Re-creating paths in Cairo is quite heavy, so this is in fact an optimization for Cairo. Same optimization was applied for Path::translate() for Cairo platforms.

nikolaszimmermann replied:
Regarding the comment from @heycam - I think the identity check should be at the call-site, but that's my personal preference, one can potentially combine multiple such checks into one at the call-sites, where you cannot do that if the called function is not inlined...

nikolaszimmermann commented on Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
```diff
@@ -87,12 +87,11 @@ void LegacyRenderSVGShape::fillShape(GraphicsContext& context) const
-        usePath = nonScalingStrokePath(usePath, nonScalingStrokeTransform());
-
-    context.strokePath(*usePath);
+        context.strokePath(nonScalingStrokePath(path()));
```
Not sure if the nonScalingStrokePath() helper is worth it, since you need access to both the Path and the AffineTransform nonScalingTransform, in most places.

```
context.strokePath(nonScalingStrokeTransform().mapPath(path()));
```

seems quite readable as-is.

nikolaszimmermann commented on Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
```diff
@@ -101,8 +100,8 @@ bool LegacyRenderSVGShape::shapeDependentStrokeContains(const FloatPoint& point,
 
     if (hasNonScalingStroke() && pointCoordinateSpace != LocalCoordinateSpace) {
         AffineTransform nonScalingTransform = nonScalingStrokeTransform();
```
s/AffineTransform/auto/ while you're at it?

nikolaszimmermann commented on Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
```diff
@@ -174,24 +173,19 @@ void LegacyRenderSVGShape::layout()
-Path* LegacyRenderSVGShape::nonScalingStrokePath(const Path* path, const AffineTransform& strokeTransform) const
+Path LegacyRenderSVGShape::nonScalingStrokePath(const Path& path) const
 {
-    static NeverDestroyed<Path> tempPath;
```
Ugh, good to see this go. This predates, Path-is-cheaply-movable times.

nikolaszimmermann commented on Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
```diff
@@ -174,24 +173,19 @@ void LegacyRenderSVGShape::layout()
-    tempPath.get().transform(strokeTransform);
-
-    return &tempPath.get();
+    return nonScalingStrokeTransform().mapPath(path);
```
.. however this is simple enough that the function call overhead could be saved by omitting the helper function altogether.

nikolaszimmermann commented on Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
```diff
@@ -174,24 +173,19 @@ void LegacyRenderSVGShape::layout()
+bool LegacyRenderSVGShape::setupNonScalingStrokeContext(GraphicsContextStateSaver& stateSaver)
 {
-    std::optional<AffineTransform> inverse = strokeTransform.inverse();
+    std::optional<AffineTransform> inverse = nonScalingStrokeTransform().inverse();
```
```
if (auto inverse = nonScalingStrokeTransform().inverse()) {
    stateSaver.save();
    stateSaver.context()->concatCTM(inverse.value());
    return true;
}

return false;
```

is more readable to my eyes.

nikolaszimmermann commented on Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
```diff
@@ -254,8 +248,7 @@ void LegacyRenderSVGShape::strokeShape(GraphicsContext& context)
     if (hasNonScalingStroke()) {
-        AffineTransform nonScalingTransform = nonScalingStrokeTransform();
-        if (!setupNonScalingStrokeContext(nonScalingTransform, stateSaver))
+        if (!setupNonScalingStrokeContext(stateSaver))
```
You could merge both if conditions.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGPath.cpp:
```diff
@@ -95,10 +94,10 @@ void RenderSVGPath::strokeShape(GraphicsContext& context) const
     useStrokeStyleToFill(context);
     for (size_t i = 0; i < m_zeroLengthLinecapLocations.size(); ++i) {
-        usePath = zeroLengthLinecapPath(m_zeroLengthLinecapLocations[i]);
+        Path usePath = zeroLengthLinecapPath(m_zeroLengthLinecapLocations[i]);
```
s/Path/auto/

nikolaszimmermann replied:
Won't repeat the other comments already written for LegacyRenderSVGShape, since it's the same code.

nikolaszimmermann commented on Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
```diff
@@ -101,8 +100,8 @@ bool LegacyRenderSVGShape::shapeDependentStrokeContains(const FloatPoint& point,
-        Path* usePath = nonScalingStrokePath(m_path.get(), nonScalingTransform);
-        return usePath->strokeContains(nonScalingTransform.mapPoint(point), [this] (GraphicsContext& context) {
+        Path usePath = nonScalingTransform.mapPath(path());
+        return usePath.strokeContains(nonScalingTransform.mapPoint(point), [this] (GraphicsContext& context) {
```
The lambda on the other hand is repeated four times as-is, and that deserves a helper to avoid the duplication.

nikolaszimmermann reviewed (commented):
Thanks @heycam, happy to give an unofficial review. Code looks good, except for a few nitpicks.

heycam committed:
Simplify non-scaling-stroke handling in (Legacy)RenderSVGShape
https://bugs.webkit.org/show_bug.cgi?id=241754

Reviewed by NOBODY (OOPS!).

(Legacy)RenderSVGShape::nonScalingStrokePath has a static variable that
stores the returned Path, presumably so that it can live long enough for
the caller to make use of it. We can simplify this by returning Path
object directly.

While we're here, we can also simplify some other functions dealing with
the non-scaling stroke transform and usage of the non-scaling stroke
path.

We add a new AffineTransform::mapPath function, which does what the
current nonScalingStrokePath function does.

* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::Path::transform):
* Source/WebCore/platform/graphics/transforms/AffineTransform.cpp:
(WebCore::AffineTransform::mapPath const):
* Source/WebCore/platform/graphics/transforms/AffineTransform.h:
* Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::strokeShape const):
(WebCore::LegacyRenderSVGShape::shapeDependentStrokeContains):
(WebCore::LegacyRenderSVGShape::setupNonScalingStrokeContext):
(WebCore::LegacyRenderSVGShape::strokeShape):
(WebCore::LegacyRenderSVGShape::calculateStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::nonScalingStrokePath const): Deleted.
* Source/WebCore/rendering/svg/LegacyRenderSVGShape.h:
(WebCore::LegacyRenderSVGShape::strokeStyleApplier const):
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::strokeShape const):
(WebCore::RenderSVGPath::zeroLengthLinecapPath const):
* Source/WebCore/rendering/svg/RenderSVGPath.h:
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::strokeShape const):
(WebCore::RenderSVGShape::shapeDependentStrokeContains):
(WebCore::RenderSVGShape::setupNonScalingStrokeContext):
(WebCore::RenderSVGShape::strokeShape):
(WebCore::RenderSVGShape::calculateStrokeBoundingBox const):
(WebCore::RenderSVGShape::nonScalingStrokePath const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGShape.h:
(WebCore::RenderSVGShape::strokeStyleApplier const):

shallawa commented on Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
```diff
@@ -253,11 +238,8 @@ void LegacyRenderSVGShape::strokeShape(GraphicsContext& context)
-        if (!setupNonScalingStrokeContext(nonScalingTransform, stateSaver))
-            return;
-    }
+    if (hasNonScalingStroke() && !setupNonScalingStrokeContext(stateSaver))
```
`nonScalingStrokeTransform` is always passed when calling `setupNonScalingStrokeContext()`. Similarly `hasNonScalingStroke()` is checked before calling it. So I would suggest moving `if (hasNonScalingStroke())` to `setupNonScalingStrokeContext()`.

heycam replied:
Confused by this comment. Are you saying I should rewrite:

```c++
if (hasNonScalingStroke() && !setupNonScalingStrokeContext(...))
    return;
```

to

```c++
if (hasNonScalingStroke()) {
    if (!setupNonScalingStrokeContext(...)
        return;
}
```

?

shallawa replied:
Sorry for the confusion. I fixed the comment. I meant adding the following lines to `setupNonScalingStrokeContext()`:

```
bool LegacyRenderSVGShape::setupNonScalingStrokeContext(GraphicsContextStateSaver& stateSaver)
{
    if (hasNonScalingStroke())
        return true;
    AffineTransform nonScalingTransform = nonScalingStrokeTransform();
    ...
}

```

And instead of calling it like this:

```
    if (hasNonScalingStroke()) {
        AffineTransform nonScalingTransform = nonScalingStrokeTransform();
        if (!setupNonScalingStrokeContext(nonScalingTransform, stateSaver))
            return;
    }
```

It can be just

```
    if (!setupNonScalingStrokeContext(stateSaver))
        return;

```

shallawa commented on Source/WebCore/platform/graphics/transforms/AffineTransform.cpp:
```diff
@@ -327,6 +328,19 @@ Region AffineTransform::mapRegion(const Region& region) const
+{
+    if (isIdentityOrTranslation()) {
+        Path mappedPath(path);
+        mappedPath.translate(roundedIntSize(FloatSize(narrowPrecisionToFloat(m_transform[4]), narrowPrecisionToFloat(m_transform[5]))));
```
Path::translate() takes FloatSize. Is there a need to create a FloatSize, rounded to IntSize and then implicitly cast it to FloatSize?

shallawa commented on Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
```diff
@@ -174,25 +169,15 @@ void LegacyRenderSVGShape::layout()
-    std::optional<AffineTransform> inverse = strokeTransform.inverse();
-    if (!inverse)
-        return false;
+    if (auto inverse = nonScalingStrokeTransform().inverse()) {
```
I think the early return `if (!inverse)` looks clearer.
